python/semantic_kernel/planning/stepwise_planner/stepwise_planner.py

```diff
@@ -0,0 +1,461 @@
+
+        if final_answer_match:
+            result.final_answer = final_answer_match.group(1).strip()
+            return result
+
+        # Extract thought
+        thought_match = re.search(S_THOUGHT_REGEX, input)
+
+        if thought_match:
+            result.thought = thought_match.group(0).strip()
+        elif not input.contains(ACTION):
+            result.thought = input
+        else:
+            raise ValueError("Unexpected input format")
+
+        result.thought = result.thought.replace(THOUGHT, "").strip()
+
+        # Extract action
+        action_match = re.search(S_ACTION_REGEX, input)
+
+        if action_match:
+            action_json = action_match.group(1).strip()
+
+            try:
+                system_step_results = json.loads(action_json)
+
+                if system_step_results is None or len(system_step_results) == 0:
+                    result.observation = (
+                        f"System step parsing error, empty JSON: {action_json}"
+                    )
+                else:
+                    result.action = system_step_results["action"]
+                    result.action_variables = system_step_results["action_variables"]
+            except Exception:
+                result.observation = (
+                    f"System step parsing error, invalid JSON: {action_json}"
+                )
+
+        if is_null_or_empty(result.thought) and is_null_or_empty(result.action):
+            result.observation = (
+                "System step error, no thought or action found.",
+                "Please give a valid thought and/or action.",
+            )
+
+        return result
+
+    def add_execution_stats_to_context(
+        self, steps_taken: List[SystemStep], context: SKContext
+    ):
+        context.variables.set("step_count", str(len(steps_taken)))
+        context.variables.set(
+            "steps_taken", json.dumps([s.__dict__ for s in steps_taken], indent=4)
+        )
+
+        action_counts: Dict[str, int] = {}
+        for step in steps_taken:
+            if is_null_or_empty(step.action):
+                continue
+
+            current_count = action_counts.get(step.action, 0)
+            action_counts[step.action] = current_count + 1
+
+        skill_call_list_with_counts = [
+            f"{skill}({action_counts[skill]})" for skill in action_counts
+        ]
+        skill_call_list_with_counts = ", ".join(skill_call_list_with_counts)
+        skill_call_count_str = str(sum(action_counts.values()))
+
+        context.variables.set(
+            "skill_count", f"{skill_call_count_str} ({skill_call_list_with_counts})"
+        )
+
+    def create_scratch_pad(self, question: str, steps_taken: List[SystemStep]) -> str:
+        if len(steps_taken) == 0:
+            return ""
+
+        scratch_pad_lines: List[str] = []
+
+        # Add the original first thought
+        # joowon-dm-snu: does prompt really need SCRATCH_PAD_PREFIX? it looks like it breaks thought process pattern
```

Comment:
I don't think there's a right answer, it's an engineering thing and it's probably going to feel different to different people. I'm going to delete it because I'm just making a note to check it out in another branch.

Reply:
Personally, I would prefer to remove the prefix, as I think maintaining the pattern helps the behavior of the model.

Reply:
We should make sure we have a set of tests that we can use to validate performance and reliability of the planners against changes.

Reply:
@lemillermicrosoft 
I completely agree. Does you team have any plans to provide such functionalities within the semantic-kernel in the future? Or plan to separate them into a different repository, similar to `semantic-memory`?

Reply:
In short, the kernel syntax examples project is something we'd like to transform into integration tests.

Reply:
I understand the concern about breaking the pattern, the scope however is only when MaxTokens limitations are hit and we need to truncate the thought process. I've updated to `... I've removed some of my previous work to make room for the new stuff ...` at the moment and have had satisfactory results with that (and other changes, see draft: #2504 and it's cousin #2444). I wouldn't worry about it too much right now though, we can iterate and improve.

Reply:
@lemillermicrosoft Agree with MaxTokens limitations! I'll use the planner as much as possible for my projects. and will try to give you some good feedback on to improve it :)